File: backend/app/services/rag.py
from __future__ import annotations
from functools import lru_cache
from pathlib import Path
from typing import Iterable
import hashlib
import chromadb
from chromadb.utils.embedding_functions import OpenAIEmbeddingFunction, SentenceTransformerEmbeddingFunction
from pypdf import PdfReader
from app.core.config import get_settings

settings = get_settings()

ROLE_COLLECTIONS = {
    "AI/ML Engineer": "kb_ai_ml",
    "Backend Engineer": "kb_backend",
    "Data Science / Applied ML": "kb_data_science",
}

ROLE_DIRECTORIES = {
    "AI/ML Engineer": "AI_ML_Engineer",
    "Backend Engineer": "Backend_Engineer",
    "Data Science / Applied ML": "Data_Science_Applied_ML",
}
KNOWN_ROLE_DIRECTORY_NAMES = set(ROLE_DIRECTORIES.values())

Path(settings.chroma_dir).mkdir(parents=True, exist_ok=True)
_CHROMA_CLIENT = chromadb.PersistentClient(path=settings.chroma_dir)


class EmbeddingConfigurationError(RuntimeError):
    pass


def _model_slug(value: str) -> str:
    return value.lower().replace("/", "_").replace("-", "_").replace(".", "_").replace(" ", "_")


@lru_cache
def get_embedder():
    provider = settings.embedding_provider.lower().strip()
    if provider == "openai":
        if not settings.openai_api_key:
            raise EmbeddingConfigurationError(
                "OPENAI_API_KEY is required when EMBEDDING_PROVIDER=openai"
            )
        return OpenAIEmbeddingFunction(
            api_key=settings.openai_api_key,
            model_name=settings.embedding_model,
        )
    if provider in {"sentence-transformer", "sentence_transformer", "local"}:
        try:
            return SentenceTransformerEmbeddingFunction(model_name=settings.embedding_model)
        except Exception as exc:
            raise EmbeddingConfigurationError(
                "Local embedding model could not be initialized. On first run, ensure internet access is available "
                "to download the sentence-transformer model or pre-cache it locally."
            ) from exc
    raise EmbeddingConfigurationError(
        f"Unsupported embedding provider: {settings.embedding_provider}"
    )

def read_document(path: Path) -> str:
    if path.suffix.lower() == ".pdf":
        reader = PdfReader(str(path))
        return "\n".join(page.extract_text() or "" for page in reader.pages)
    return path.read_text(encoding="utf-8", errors="ignore")


def role_directory_name(role: str) -> str:
    return ROLE_DIRECTORIES.get(role, role.replace("/", "_").strip())


def role_from_directory_name(directory_name: str) -> str:
    for role, mapped_name in ROLE_DIRECTORIES.items():
        if mapped_name == directory_name:
            return role
    return directory_name.replace("_", " ")


def collection_for_role(role: str):
    base_name = ROLE_COLLECTIONS.get(role, f"kb_{role.lower().replace(' ', '_').replace('/', '_')}")
    provider_slug = _model_slug(settings.embedding_provider)
    model_slug = _model_slug(settings.embedding_model)
    collection_name = f"{base_name}_{provider_slug}_{model_slug}"
    return _CHROMA_CLIENT.get_or_create_collection(collection_name, embedding_function=get_embedder())

def chunk_text(text: str, chunk_size: int = 950, overlap: int = 160) -> list[str]:
    words = text.split()
    chunks, start = [], 0
    while start < len(words):
        end = min(start + chunk_size, len(words))
        chunks.append(" ".join(words[start:end]))
        if end == len(words):
            break
        start = max(0, end - overlap)
    return [c for c in chunks if len(c) > 120]

def ingest_knowledge_base(role: str | None = None) -> dict:
    kb_dir = Path(settings.kb_dir)
    kb_dir.mkdir(parents=True, exist_ok=True)
    if role:
        role_dirs = [(role, kb_dir / role_directory_name(role))]
    else:
        role_dirs = [
            (role_from_directory_name(path.name), path)
            for path in kb_dir.iterdir()
            if path.is_dir() and path.name in KNOWN_ROLE_DIRECTORY_NAMES
        ]
    total = 0
    scanned_roles = []
    for role_name, rdir in role_dirs:
        if not rdir.exists():
            continue
        scanned_roles.append(role_name)
        col = collection_for_role(role_name)
        for doc in list(rdir.glob("*.pdf")) + list(rdir.glob("*.txt")) + list(rdir.glob("*.md")):
            chunks = chunk_text(read_document(doc))
            ids = [hashlib.sha1(f"{doc}:{i}:{c[:80]}".encode()).hexdigest() for i, c in enumerate(chunks)]
            if chunks:
                col.upsert(ids=ids, documents=chunks, metadatas=[{"source": doc.name, "role": role_name, "chunk": i} for i in range(len(chunks))])
                total += len(chunks)
    return {"chunks_ingested": total, "roles_scanned": scanned_roles}


def build_queries(role: str, resume_profile: dict, last_answer: str | None = None) -> list[str]:
    skills = resume_profile.get("skills", [])
    projects = resume_profile.get("projects", [])
    experience = resume_profile.get("years_experience") or resume_profile.get("seniority_signal") or "entry-level"
    domains = resume_profile.get("domains", [])
    role_topics = {
        "AI/ML Engineer": "modeling, data pipelines, evaluation, experimentation, and deployment",
        "Backend Engineer": "apis, databases, reliability, debugging, and service design",
        "Data Science / Applied ML": "analysis, experimentation, feature engineering, and model delivery",
    }
    role_topic = role_topics.get(role, "practical role-specific engineering concepts")

    queries = []

    if skills:
        queries.append(f"{role} interview concepts for {', '.join(skills[:5])}")
        queries.append(f"practical debugging and design scenarios using {', '.join(skills[:4])}")

    if projects:
        queries.append(f"{role} concepts related to candidate project: {projects[0]}")

    if domains:
        queries.append(f"{role} interview topics covering {', '.join(domains[:3])}")

    queries.append(f"{role} {experience} interview topics covering {role_topic}")

    if last_answer:
        queries.append(f"follow up question based on candidate answer: {last_answer[:300]}")

    return queries

def retrieve(role: str, queries: Iterable[str], k: int = 4) -> list[dict]:
    col = collection_for_role(role)
    results = []
    for q in queries:
        res = col.query(query_texts=[q], n_results=k)
        for doc, meta, dist in zip(res.get("documents", [[]])[0], res.get("metadatas", [[]])[0], res.get("distances", [[]])[0]):
            results.append({"text": doc[:1200], "metadata": meta, "distance": dist, "query": q})
    seen, dedup = set(), []
    for r in results:
        key = (r["metadata"].get("source"), r["metadata"].get("chunk"))
        if key not in seen:
            dedup.append(r)
            seen.add(key)
    return dedup[:k]

File: netlify/functions/_lib/rag.ts
import { appConfig } from './config.js';
import { HttpError } from './errors.js';
import { embedTexts } from './openai.js';
import { extractTextFromUpload } from './resume.js';
import { listFolder, downloadBuffer } from './storage.js';
import { supabase } from './supabase.js';
import type { ResumeProfile, RetrievedContext } from './types.js';

export const ROLE_DIRECTORIES: Record<string, string> = {
  'AI/ML Engineer': 'AI_ML_Engineer',
  'Backend Engineer': 'Backend_Engineer',
  'Data Science / Applied ML': 'Data_Science_Applied_ML',
};

const ROLE_COLLECTIONS: Record<string, string> = {
  'AI/ML Engineer': 'modeling, data pipelines, evaluation, experimentation, and deployment',
  'Backend Engineer': 'apis, databases, reliability, debugging, and service design',
  'Data Science / Applied ML': 'analysis, experimentation, feature engineering, and model delivery',
};

export function roleDirectoryName(role: string) {
  return ROLE_DIRECTORIES[role] || role.replace(/\//g, '_').trim();
}

function chunkText(text: string, chunkSize = 950, overlap = 160) {
  const words = text.split(/\s+/).filter(Boolean);
  const chunks: string[] = [];
  let start = 0;
  while (start < words.length) {
    const end = Math.min(start + chunkSize, words.length);
    const chunk = words.slice(start, end).join(' ');
    if (chunk.length > 120) chunks.push(chunk);
    if (end === words.length) break;
    start = Math.max(0, end - overlap);
  }
  return chunks;
}

export async function ingestKnowledgeBase(role?: string) {
  const roles = role ? [role] : Object.keys(ROLE_DIRECTORIES);
  let total = 0;
  const scannedRoles: string[] = [];

  for (const roleName of roles) {
    const directory = roleDirectoryName(roleName);
    const files = await listFolder(appConfig.knowledgeBaseBucket, directory);
    if (!files.length) continue;
    scannedRoles.push(roleName);

    for (const file of files) {
      if (!/\.(pdf|txt|md)$/i.test(file.name)) continue;
      const storagePath = `${directory}/${file.name}`;
      const bytes = await downloadBuffer(appConfig.knowledgeBaseBucket, storagePath);
      const text = await extractTextFromUpload(file.name, bytes);
      const chunks = chunkText(text);
      if (!chunks.length) continue;
      const embeddings = await embedTexts(chunks);

      const { error: deleteError } = await supabase.from('kb_chunks').delete().eq('role', roleName).eq('source', file.name);
      if (deleteError) {
        throw new HttpError(500, `Failed to clear old chunks for ${file.name}`, deleteError);
      }

      const rows = chunks.map((content, index) => ({
        role: roleName,
        source: file.name,
        storage_path: storagePath,
        chunk_index: index,
        content,
        metadata: {
          source: file.name,
          role: roleName,
          chunk: index,
        },
        embedding: embeddings[index],
      }));

      const { error: insertError } = await supabase.from('kb_chunks').insert(rows);
      if (insertError) {
        throw new HttpError(500, `Failed to store chunks for ${file.name}`, insertError);
      }
      total += rows.length;
    }
  }

  return { chunks_ingested: total, roles_scanned: scannedRoles };
}

export function buildQueries(role: string, resumeProfile: ResumeProfile, lastAnswer?: string) {
  const skills = resumeProfile.skills || [];
  const projects = resumeProfile.projects || [];
  const experience = resumeProfile.years_experience || resumeProfile.seniority_signal || 'entry-level';
  const domains = resumeProfile.domains || [];
  const roleTopic = ROLE_COLLECTIONS[role] || 'practical role-specific engineering concepts';

  const queries: string[] = [];
  if (skills.length) {
    queries.push(`${role} interview concepts for ${skills.slice(0, 5).join(', ')}`);
    queries.push(`practical debugging and design scenarios using ${skills.slice(0, 4).join(', ')}`);
  }
  if (projects.length) {
    queries.push(`${role} concepts related to candidate project: ${projects[0]}`);
  }
  if (domains.length) {
    queries.push(`${role} interview topics covering ${domains.slice(0, 3).join(', ')}`);
  }
  queries.push(`${role} ${experience} interview topics covering ${roleTopic}`);
  if (lastAnswer) {
    queries.push(`follow up question based on candidate answer: ${lastAnswer.slice(0, 300)}`);
  }
  return queries;
}

export async function retrieveContext(role: string, queries: string[], count = 4): Promise<RetrievedContext[]> {
  if (!queries.length) return [];
  const embeddings = await embedTexts(queries);
  const results: RetrievedContext[] = [];

  for (let index = 0; index < queries.length; index += 1) {
    const { data, error } = await supabase.rpc('match_kb_chunks', {
      filter_role: role,
      query_embedding: embeddings[index],
      match_count: count,
    });
    if (error) {
      throw new HttpError(500, 'Failed to retrieve knowledge base context', error);
    }
    for (const row of data || []) {
      results.push({
        text: row.content.slice(0, 1200),
        metadata: row.metadata || { source: row.source, chunk: row.chunk_index, role: row.role },
        similarity: row.similarity,
        query: queries[index],
      });
    }
  }

  const deduped: RetrievedContext[] = [];
  const seen = new Set<string>();
  for (const item of results) {
    const key = `${item.metadata.source ?? 'source'}:${item.metadata.chunk ?? 'chunk'}`;
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(item);
    if (deduped.length >= count) break;
  }
  return deduped;
}
